qiskit/circuit/quantumcircuit.py

```diff
@@ -2046,10 +2070,10 @@ def copy(self, name: str | None = None) -> "QuantumCircuit":
           QuantumCircuit: a deepcopy of the current circuit, with the specified name
         """
         cpy = self.copy_empty_like(name)
+        cpy._data = self._data.copy()
 
-        operation_copies = {
-            id(instruction.operation): instruction.operation.copy() for instruction in self._data
-        }
+        operation_copies = {}
+        self._data.foreach_op(lambda op: operation_copies.update({id(op): op.copy()}))
```

Comment:
```suggestion
        def memo_copy(op):
            if (out := operation_copies.get(id(op))) is not None:
                return out
            return (operation_copies[id(op)] := op.copy())
        self._data.replace_ops(memo_copy)
```
or something like that, maybe? Better to avoid the temporary allocations of loads of one-item dicts, and this does the replacement in the same loop as the memoisation.

Reply:
Done in b68a6aa0bb7dd3eb2453baeaf6db6a1cd5612dc5